test/metabase/models/card_test.clj

```diff
@@ -1173,3 +1175,154 @@
+  (mt/with-temp [:model/Card {id :id} {:name          "A card"
+                                       :dataset_query (bare-query)}]
+    ;; :idents are populated on initial insert; update to remove them. (Update does not populate them like insert.)
+    (t2/update! :model/Card id {:dataset_query (bare-query)})
+    ;; Can't use the one from `with-temp` since it came before the above edit.
+    (let [{eid   :entity_id
+           query :dataset_query} (t2/select-one :model/Card :id id)]
+      (testing "on read, a :dataset_query with missing idents gets them filled in based on entity_id"
+        ;; These idents are: kind_EID@stage__index, eg. "aggregation_4QsLuEnriHKkXCWqbPMQ8@0__0"
+        (is (string? eid))
+        (is (=? (bare-query-exp eid)
+                (:query query)))))))
+
+(deftest ^:sequential fall-back-to-hashing-entity-id-test
+  (mt/with-temp [:model/Card {id :id} {:name          "A card"
+                                       :dataset_query (bare-query)}]
+    ;; :idents are populated on initial insert; update to remove them. (Update does not populate them like insert.)
+    ;; Also remove the generated :entity_id.
+    (t2/update! :model/Card id {:dataset_query (bare-query)
+                                :entity_id     nil})
+    ;; Can't use the one from `with-temp` since it came before the above edit.
+    (let [{eid   :entity_id
+           query :dataset_query} (t2/select-one :model/Card :id id)]
+      (testing "on read, a :dataset_query with missing idents AND :entity_id gets a hashed entity_id and idents"
+        ;; These idents are: kind_EID@stage__index, eg. "aggregation_4QsLuEnriHKkXCWqbPMQ8@0__0"
+        ;; The entity_id is hashed based on created_at, so it's still always different!
+        (is (string? eid))
+        (is (=? (bare-query-exp eid)
+                (:query query)))))))
+
+(deftest ^:sequential e2e-entity-id-and-idents-test
+  (mt/with-temp [:model/Card {id :id} {:name          "A card"
+                                       :dataset_query (bare-query)}]
+    ;; :idents are populated on initial insert; update to remove them. (Update does not populate them like insert.)
+    ;; Also remove the generated :entity_id.
+    (t2/update! :model/Card id {:dataset_query (bare-query)
+                                :entity_id     nil})
+    ;; Can't use the one from `with-temp` since it came before the above edit.
+    (let [{eid   :entity_id
+           query :dataset_query} (t2/select-one :model/Card :id id)]
+      (testing "idents should be populated on read"
+        ;; These idents are: kind_EID@stage__index, eg. "aggregation_4QsLuEnriHKkXCWqbPMQ8@0__0"
+        ;; The entity_id is hashed based on created_at, so it's still always different!
+        (is (string? eid))
+        (is (=? (bare-query-exp eid)
+                (:query query)))
+
+        (testing "but not written back to appdb"
+          (let [{raw-query :dataset_query} (t2/select-one "report_card" :id id)]
```

Comment:
```suggestion
          (let [{raw-query :dataset_query} (t2/select-one :report_card :id id)]
```

Reply:
I thought that keywords were for models and strings for raw table names?

It seems to work with a keyword, but is there an idiom here?

Reply:
I'm not sure, I've always used keywords.

Reply:
Well it works, so I'll run with that.